README
This MariaDB plugin authenticates database users against the system
password database or other authentication mechanisms supported by
saslauthd.

ENABLE SASLAUTHD

On Slackware Linux, enable the saslauthd daemon with:

  chmod +x /etc/rc.d/rc.saslauthd
  /etc/rc.d/rc.saslauthd start

On OpenBSD, install the cyrus-sasl package and enable the saslauthd daemon
with rcctl(8).

On Debian, install the sasl2-bin package and enable the saslauthd daemon. Add
the mysql user to the sasl group and restart the database system. Create the
file /etc/pam.d/mariadb. Example:

  #%PAM-1.0
  @include common-auth
  @include common-account

PLUGIN INSTALLATION

To install the plugin, run the following commands:

  make OPTIMIZE="-O2" SASLAUTHD_PATH="/var/run/saslauthd/mux"
  make install

Depending on your system the communications socket is named
/run/saslauthd/mux, /var/run/saslauthd/mux, /var/state/saslauthd/mux or
/var/sasl2/mux.

Enable the plugin with:

  mysql -h localhost -u root -p \
        -e "INSTALL PLUGIN saslauthd SONAME 'auth_saslauthd.so'"

Or add the following section to your database server's configuration and
restart the database system:

  [mariadb]
  plugin-load-add=auth_saslauthd.so

List the installed plugins with:

  mysql -h localhost -u root -p -e 'SHOW PLUGINS\G'

     Name: saslauthd
   Status: ACTIVE
     Type: AUTHENTICATION
  Library: auth_saslauthd.so
  License: BSD

DATABASE USERS

Add a database account for the system user "jennifer" as follows:

  CREATE USER 'jennifer'@'localhost' IDENTIFIED WITH saslauthd;

Optionally, a different system user as well as a realm can be specified.
Example:

  CREATE USER 'jekyll' IDENTIFIED WITH saslauthd AS 'hyde@EXAMPLE.COM';

Run the mysql command-line tool and enter the system user's password to
check the database account. Example:

  mysql -h localhost -u jennifer -p
  Enter password:

SECURE CONNECTIONS

saslauthd needs clear text passwords. Secure non-local connections
between your database clients and the server with TLS. See the MariaDB
documentation for more information.

DEPENDENCIES

The plugin requires MariaDB and the saslauthd daemon from Cyrus SASL.

Building the plugin requires GNU Make 4 or a BSD make.

LICENSE AND COPYRIGHT

Copyright 2017 Andreas Voegele

This plugin is free software; you can redistribute it and/or modify it
under the terms of the ISC license.


Changes
Revision history for mariadb-plugin-saslauthd

1.5  2018-08-07 05:29:32 CEST

  - Don't leak memory on error.

1.4  2018-07-03 18:10:49 CEST

  - Use maria_declare_plugin_end.

1.3  2018-06-04 06:05:49 CEST

  - Include my_global.h instead of my_sys.h.
  - Use maria_declare_plugin if available.

1.2  2017-09-02 07:25:32 CEST

  - Add /usr/include/mysql/server to the include path.

1.1  2017-07-22 08:00:37 CEST

  - Use strerror_r() instead of strerror().
  - Add .travis.yml.

1.0  2017-07-09 16:49:48 CEST

  - First version, released on an unsuspecting world.



saslauthd_client.c
/*
 * Verify user name and password with saslauthd
 *
 * Copyright 2017-2018 Andreas Voegele <andreas@andreasvoegele.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/socket.h>
#include <sys/un.h>
#include <sys/uio.h>
#include <arpa/inet.h>

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <unistd.h>

#include "saslauthd_client.h"

static void
set_error(struct saslauthd_credentials *cred, const char *format, ...)
{
  va_list ap;

  va_start(ap, format);
  (void) vsnprintf(cred->error, sizeof(cred->error), format, ap);
  va_end(ap);
}

static ssize_t
write_string(int fd, const char *s, unsigned short len)
{
  unsigned short count;
  struct iovec iov[2];
  int iovcnt;
  ssize_t n;

  iovcnt = (s != NULL) ? 2 : 1;
  count = htons(len);
  iov[0].iov_base = (void *) &count;
  iov[0].iov_len = sizeof(count);
  iov[1].iov_base = (void *) s;
  iov[1].iov_len = len;

  while ((n = writev(fd, iov, iovcnt)) == -1 && errno == EINTR)
    ;

  return n;
}

static ssize_t
read_string(int fd, char **ps)
{
  unsigned short count;
  char *s;
  ssize_t n;

  while ((n = read(fd, &count, sizeof(count))) == -1 && errno == EINTR)
    ;
  if (n < (ssize_t) sizeof(count))
    return -1;

  count = ntohs(count);

  s = (char *) malloc(count + 1);
  if (s == NULL)
    return -1;

  while ((n = read(fd, s, count)) == -1 && errno == EINTR)
    ;
  if (n < (ssize_t) count)
  {
    free(s);
    return -1;
  }

  s[count] = '\0';
  *ps = s;

  return n;
}

int
saslauthd_checkpass(const char *path, struct saslauthd_credentials *cred)
{
  int fd = -1;
  struct sockaddr_un sun;
  char *reply = NULL;
  ssize_t reply_len;
  int rc = -1;

  if ((fd = socket(AF_UNIX, SOCK_STREAM, 0)) < 0)
  {
    char error[128];
    strerror_r(errno, error, sizeof(error));
    set_error(cred, "Cannot create socket: %s", error);
    goto end;
  }

  memset(&sun, 0, sizeof(sun));
  sun.sun_family = AF_UNIX;
  strncpy(sun.sun_path, path, sizeof(sun.sun_path) - 1);
  if (connect(fd, (struct sockaddr *) &sun, sizeof(sun)) < 0)
  {
    char error[128];
    strerror_r(errno, error, sizeof(error));
    set_error(cred, "Cannot connect to '%s': %s", path, error);
    goto end;
  }

  if (write_string(fd, cred->user, cred->user_len) < 0)
  {
    set_error(cred, "Failed to send userid");
    goto end;
  }

  if (write_string(fd, cred->password, cred->password_len) < 0)
  {
    set_error(cred, "Failed to send password");
    goto end;
  }

  if (write_string(fd, cred->service, cred->service_len) < 0)
  {
    set_error(cred, "Failed to send service name");
    goto end;
  }

  if (write_string(fd, cred->realm, cred->realm_len) < 0)
  {
    set_error(cred, "Failed to send realm");
    goto end;
  }

  if ((reply_len = read_string(fd, &reply)) < 2)
  {
    set_error(cred, "Received invalid reply");
    goto end;
  }

  if (strncmp(reply, "OK", 2) == 0)
    rc = 0;
  else
  {
    const char *reason = "Unknown reason";
    if (strncmp(reply, "NO", 2) == 0)
    {
      rc = 1;
      if (reply_len > 3)
        reason = &reply[3];
    }
    set_error(cred, "%s", reason);
  }

end:
  if (reply != NULL)
    free(reply);
  if (fd >= 0)
    (void) close(fd);

  return rc;
}


saslauthd_client.h
/*
 * Verify user name and password with saslauthd
 *
 * Copyright 2017-2018 Andreas Voegele <andreas@andreasvoegele.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

struct saslauthd_credentials
{
  const char *user;
  unsigned short user_len;

  const char *password;
  unsigned short password_len;

  const char *service;
  unsigned short service_len;

  const char *realm;
  unsigned short realm_len;

  char error[256];
};

/**
 * Verifies a user name and password with saslauthd.
 *
 * @param  path the saslauthd communications socket
 * @param  cred a filled saslauthd_credentials object
 * @return      0 on success, 1 on authentication failure and -1 on error
 */
extern int saslauthd_checkpass(const char *, struct saslauthd_credentials *);


auth_saslauthd.c
/*
 * Verify user name and password with saslauthd
 *
 * Copyright 2017-2018 Andreas Voegele <andreas@andreasvoegele.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <stdlib.h>
#include <string.h>

#include <mysql/plugin_auth.h>
#include <my_global.h>
#include <mysqld_error.h>
#include <mysql_version.h>

#include "saslauthd_client.h"

#ifndef SASLAUTHD_PATH
#define SASLAUTHD_PATH "/var/run/saslauthd/mux"
#endif

#ifndef SASLAUTHD_SERVICE
#define SASLAUTHD_SERVICE "mariadb"
#endif

static char *saslauthd_path;
static char *saslauthd_service;

static int
get_user(MYSQL_SERVER_AUTH_INFO *info, struct saslauthd_credentials *cred,
         char **pbuf)
{
  const char *user, *realm, *extuser;
  unsigned short user_len, realm_len, extuser_len;
  char *buf;

  user = info->user_name;
  user_len = info->user_name_length;
  realm = NULL;
  realm_len = 0;
  buf = NULL;

  extuser = user;
  extuser_len = user_len;

  if (info->auth_string != NULL && info->auth_string_length > 0)
  {
    char *sep;
    size_t buf_len;

    extuser = info->auth_string;
    extuser_len = info->auth_string_length;

    buf_len = info->auth_string_length;
    buf = (char *) malloc(buf_len + 1);
    if (buf == NULL)
      return 0;
    memcpy(buf, info->auth_string, buf_len);
    buf[buf_len] = '\0';

    sep = strrchr(buf, '@');
    if (sep == NULL)
    {
      user = buf;
      user_len = buf_len;
    }
    else
    {
      sep[0] = '\0';
      realm = &sep[1];
      realm_len = buf_len - (realm - buf);
      if (sep > buf)
      {
        user = buf;
        user_len = buf_len - realm_len - 1;
      }
    }
  }

  if (extuser != NULL)
  {
    size_t max_len = sizeof(info->external_user) - 1;
    size_t len = (extuser_len > max_len) ? max_len : extuser_len;
    strncpy(info->external_user, extuser, len);
    info->external_user[len] = '\0';
  }

  cred->user = user;
  cred->user_len = user_len;
  cred->realm = realm;
  cred->realm_len = realm_len;
  *pbuf = buf;

  return 1;
}

static int
saslauthd_auth(MYSQL_PLUGIN_VIO *vio, MYSQL_SERVER_AUTH_INFO *info)
{
  unsigned char *pkt;
  int pkt_len;
  struct saslauthd_credentials cred[1];
  int rc = -1;
  char *buf = NULL;

  memset(cred, 0, sizeof(struct saslauthd_credentials));
  cred->service = saslauthd_service;
  cred->service_len = (saslauthd_service != NULL) ? strlen(saslauthd_service) : 0;
  if (!get_user(info, cred, &buf))
    goto end;

  if (vio->write_packet(vio, (const unsigned char *) "\4Enter password:", 16))
    goto end;

  if ((pkt_len = vio->read_packet(vio, &pkt)) < 0)
    goto end;

  info->password_used = PASSWORD_USED_YES;

  cred->password = (const char *) pkt;
  cred->password_len = pkt_len;
  rc = saslauthd_checkpass(saslauthd_path, cred);
  if (rc != 0)
    my_printf_error(ER_ACCESS_DENIED_ERROR, "saslauthd: %s", MYF(0),
		    cred->error);
end:
  if (buf != NULL)
    free(buf);
  return (rc == 0) ? CR_OK : CR_ERROR;
}

static MYSQL_SYSVAR_STR(path, saslauthd_path,
                        PLUGIN_VAR_RQCMDARG | PLUGIN_VAR_READONLY,
                        "saslauthd communications socket",
                        NULL,
                        NULL,
                        SASLAUTHD_PATH);

static MYSQL_SYSVAR_STR(service, saslauthd_service,
                        PLUGIN_VAR_RQCMDARG | PLUGIN_VAR_READONLY,
                        "saslauthd service name",
                        NULL,
                        NULL,
                        SASLAUTHD_SERVICE);

static struct st_mysql_sys_var *vars[] =
{
  MYSQL_SYSVAR(path),
  MYSQL_SYSVAR(service),
  NULL
};

static struct st_mysql_auth info =
{
  MYSQL_AUTHENTICATION_INTERFACE_VERSION,
  "mysql_clear_password",
  saslauthd_auth
};

#if defined(MARIADB_BASE_VERSION) && MYSQL_VERSION_ID >= 50203

maria_declare_plugin(saslauthd)
{
  MYSQL_AUTHENTICATION_PLUGIN,
  &info,
  "saslauthd",
  "Andreas Voegele",
  "saslauthd based authentication",
  PLUGIN_LICENSE_BSD,
  NULL,
  NULL,
  0x0105,
  NULL,
  vars,
  "1.5",
  MariaDB_PLUGIN_MATURITY_STABLE
}
maria_declare_plugin_end;

#else

mysql_declare_plugin(saslauthd)
{
  MYSQL_AUTHENTICATION_PLUGIN,
  &info,
  "saslauthd",
  "Andreas Voegele",
  "saslauthd based authentication",
  PLUGIN_LICENSE_BSD,
  NULL,
  NULL,
  0x0105,
  NULL,
  vars,
  NULL,
  0
}
mysql_declare_plugin_end;

#endif
